File: academic-system-java/src/main/java/org/example/academic/system/repository/XmlRepository.java
package org.example.academic.system.repository;

import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import org.example.academic.system.model.AcademicClass;
import org.example.academic.system.model.Assessment;
import org.example.academic.system.model.Assignment;
import org.example.academic.system.model.Exam;
import org.example.academic.system.model.PracticalAssignment;
import org.example.academic.system.model.Seminar;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class XmlRepository implements PersistenceStrategy {

    // Serializa a lista de turmas para o arquivo academic_data.xml usando Jackson XML
    @Override
    public void save(List<AcademicClass> classes) {
        try {
            XmlMapper xmlMapper = new XmlMapper();
            // Escreve os dados no arquivo XML formatado de forma limpa
            xmlMapper.writerWithDefaultPrettyPrinter().writeValue(new File("academic_data.xml"), toStoredClasses(classes));
        } catch (Exception e) {
            System.err.println("Erro ao salvar os dados em XML: " + e.getMessage());
        }
    }

    // Lê o arquivo XML e reconstrói a lista de turmas com suas avaliações
    @Override
    public List<AcademicClass> load() {
        File file = new File("academic_data.xml");

        // Se o arquivo ainda não existir (primeira execução), retorna a lista vazia
        if (!file.exists()) {
            return new ArrayList<>();
        }

        try {
            XmlMapper xmlMapper = new XmlMapper();
            // Lê o arquivo XML e deserializa para uma lista de AcademicClass
            List<StoredClass> storedClasses = xmlMapper.readValue(file, xmlMapper.getTypeFactory().constructCollectionType(List.class, StoredClass.class));
            List<AcademicClass> classes = toAcademicClasses(storedClasses);
            System.out.println("Dados carregados do XML com sucesso! Total de turmas: " + classes.size());
            return classes;
        } catch (Exception e) {
            System.err.println("Erro ao carregar os dados do XML: " + e.getMessage());
        }
        return new ArrayList<>();
    }

    // Retorna o nome do formato para uso nos logs e relatórios
    @Override
    public String getFormatName() {
        return "XML";
    }

    // Converte a lista de AcademicClass para DTOs simples antes de serializar em XML
    private List<StoredClass> toStoredClasses(List<AcademicClass> classes) {
        List<StoredClass> storedClasses = new ArrayList<>();

        for (AcademicClass academicClass : classes) {
            StoredClass storedClass = new StoredClass();
            storedClass.code = academicClass.getCode();
            storedClass.title = academicClass.getTitle();

            for (Assessment assessment : academicClass.getAssessments()) {
                StoredAssessment storedAssessment = new StoredAssessment();
                storedAssessment.type = assessment.getType();
                storedAssessment.value = assessment.getValue();
                storedAssessment.weight = assessment.getWeight();
                storedClass.assessments.add(storedAssessment);
            }

            storedClasses.add(storedClass);
        }

        return storedClasses;
    }

    // Reconstrói objetos AcademicClass a partir dos DTOs lidos do arquivo XML
    private List<AcademicClass> toAcademicClasses(List<StoredClass> storedClasses) {
        List<AcademicClass> classes = new ArrayList<>();

        for (StoredClass storedClass : storedClasses) {
            AcademicClass academicClass = new AcademicClass(storedClass.code, storedClass.title);

            for (StoredAssessment storedAssessment : storedClass.assessments) {
                academicClass.addAssessment(createAssessment(storedAssessment));
            }

            classes.add(academicClass);
        }

        return classes;
    }

    // Instancia a subclasse correta de Assessment com base no tipo armazenado no XML
    private Assessment createAssessment(StoredAssessment storedAssessment) {
        String type = storedAssessment.type == null ? "" : storedAssessment.type.trim().toLowerCase();

        return switch (type) {
            case "exam" -> new Exam(storedAssessment.value, storedAssessment.weight);
            case "practical", "practical assignment" -> new PracticalAssignment(storedAssessment.value, storedAssessment.weight);
            case "seminar" -> new Seminar(storedAssessment.value, storedAssessment.weight);
            case "assignment" -> new Assignment(storedAssessment.value, storedAssessment.weight);
            default -> throw new IllegalArgumentException("Tipo de avaliacao invalido: " + storedAssessment.type);
        };
    }

    public static class StoredClass {
        public String code;
        public String title;
        public List<StoredAssessment> assessments = new ArrayList<>();
    }

    public static class StoredAssessment {
        public String type;
        public double value;
        public double weight;
    }
}

File: academic-system-python/academic_system/repository/xml_repository.py
import os
import xml.etree.ElementTree as ET
from typing import List

from .persistence_strategy import PersistenceStrategy
from ..model.academic_class import AcademicClass
from ..model.assignment import Assignment
from ..model.exam import Exam
from ..model.practical_assignment import PracticalAssignment
from ..model.seminar import Seminar


class XmlRepository(PersistenceStrategy):
    FILE_NAME = "academic_data.xml"

    # Serializa a lista de turmas para o arquivo XML usando a biblioteca padrão ElementTree
    def save(self, classes: List[AcademicClass]) -> None:
        try:
            root = ET.Element("classes")
            for academic_class in classes:
                class_elem = ET.SubElement(root, "class")
                ET.SubElement(class_elem, "code").text = academic_class.code
                ET.SubElement(class_elem, "title").text = academic_class.title
                assessments_elem = ET.SubElement(class_elem, "assessments")
                for assessment in academic_class.assessments:
                    a_elem = ET.SubElement(assessments_elem, "assessment")
                    ET.SubElement(a_elem, "type").text = assessment.type
                    ET.SubElement(a_elem, "value").text = str(assessment.value)
                    ET.SubElement(a_elem, "weight").text = str(assessment.weight)
            tree = ET.ElementTree(root)
            ET.indent(tree, space="  ")
            tree.write(self.FILE_NAME, encoding="unicode", xml_declaration=False)
        except Exception as e:
            print(f"Erro ao salvar os dados em XML: {e}")

    # Lê o arquivo XML e reconstrói a lista de turmas com suas avaliações
    def load(self) -> List[AcademicClass]:
        if not os.path.exists(self.FILE_NAME):
            return []
        try:
            tree = ET.parse(self.FILE_NAME)
            root = tree.getroot()
            classes = []
            for class_elem in root.findall("class"):
                code = class_elem.findtext("code", "")
                title = class_elem.findtext("title", "")
                academic_class = AcademicClass(code, title)
                for a_elem in class_elem.findall("./assessments/assessment"):
                    t = a_elem.findtext("type", "").lower().strip()
                    value = float(a_elem.findtext("value", "0"))
                    weight = float(a_elem.findtext("weight", "0"))
                    academic_class.add_assessment(self._create_assessment(t, value, weight))
                classes.append(academic_class)
            return classes
        except Exception as e:
            print(f"Erro ao carregar os dados do XML: {e}")
            return []

    # Instancia a subclasse correta de Assessment com base no tipo lido do XML
    def _create_assessment(self, type_str: str, value: float, weight: float):
        if type_str == "exam":
            return Exam(value, weight)
        elif type_str in ("practical", "practical assignment"):
            return PracticalAssignment(value, weight)
        elif type_str == "seminar":
            return Seminar(value, weight)
        return Assignment(value, weight)

    # Retorna o nome do formato para uso nos logs e relatórios
    def get_format_name(self) -> str:
        return "XML"
